### ISSUE TITLE
Invalid translation when in v-if

### ISSUE THREAD

**Issue by mlankenau** [Polyconseil/vue-gettext#6]
I get weird results whenever I have translations in a tree where the parent is switched with v-if. It renders text, but not the right one. Looks like it renders the text directly before.

**Comment by kemar** [comment]
Can you give me an example of your code and your Vue.js version number so I can reproduce the bug?

**Comment by mlankenau** [comment]
I tried hard to create a simple example that reproduces the issue but I failed. But it is like:
```
<template>
  <div v-if="list.length > 0">
     <translate>Some headline</translate>
  </div>
  <div v-if="other_list.length > 0">
     <translate>Some other headline</translate>
  </div>
</template>
```

In that case both headlines are the same. If I change it to v-show, then it works. And I found, that when I set both arrays to null in data, then it works as well (before they were empty arrays).

**Comment by kemar** [comment]
It will be hard to debug without a proper example.

Anyway, I'll try to reproduce this behaviour as soon as I've got some free time.

**Comment by kemar** [comment]
I'm closing this issue now since I cannot reproduce your problem.

Please feel free to reopen the issue if you can provide a proper failing example :)

**Comment by lattam** [comment]
I have the same problem. Hope this snippet will help to reproduce it:
https://jsfiddle.net/lattam/gvx6s7vw/

**Comment by kemar** [comment]
Note to self: try to reproduce this in a unit test

**Comment by matthieubellon** [comment]
Hi,

Joining this ticket with also a problem concerning translation and v-if. On a radio form element I have those callout that are displayed whenever 'inviteuser.group' change.

If I remove v-translate (or <translate>) it works. If v-translate is present the callouts won't change even if the property change.

> 
                    <transition name="slide-fade">
                        <div class="callout primary" v-if="inviteuser.group === 'Administrators'" v-translate>
                            Les administrateurs ont les mêmes permissions que les éditeurs et peuvent en outre configurer le desk (créer/modifier les canaux, tags, membres du desk, ..)
                        </div>
                        <div class="callout primary" v-if="inviteuser.group === 'Editors'" v-translate>
                            Les éditeurs peuvent voir et modifier tous les contenus du desk. Ils peuvent créer et modifier des campagnes.
                        </div>
                        <div class="callout primary" v-if="inviteuser.group === 'Restricted Editors'" v-translate>
                            Les éditeurs restreints peuvent uniquement voir et modifier les contenus au sein des campagnes et canaux dont ils sont responsables.
                        </div>
                    </transition>

**Comment by kemar** [comment]
Hi @matthieubellon,

Your problem is described in the README: *Caveat when using v-translate with interpolation* in the [HTML support](https://github.com/Polyconseil/vue-gettext#html-support) section.

You can fix your problem by adding an expression to the directive to provide a changing value:

```html
<transition name="slide-fade">
  <div v-if="inviteuser.group === 'Administrators'" v-translate="inviteuser.group">
    Les administrateurs ont les mêmes permissions que les éditeurs et peuvent en outre configurer le desk (créer/modifier les canaux, tags, membres du desk, ..)
  </div>
  <div v-if="inviteuser.group === 'Editors'" v-translate="inviteuser.group">
    Les éditeurs peuvent voir et modifier tous les contenus du desk. Ils peuvent créer et modifier des campagnes.
  </div>
  <div v-if="inviteuser.group === 'Restricted Editors'" v-translate="inviteuser.group">
    Les éditeurs restreints peuvent uniquement voir et modifier les contenus au sein des campagnes et canaux dont ils sont responsables.
  </div>
</transition>
```

Concerning the other problems, I'm still unable to reproduce a single bug with `v-if`. So I'm closing this issue for now.

Feel free to submit a fully working `jsfiddle` snippet if you think there is a bug.

**Comment by matthieubellon** [comment]
Thank you @kemar :) (I've read that section but did not get it obviously)

**Comment by kennyki** [comment]
Stumbled upon this issue with `v-if` on the `v-translate` directive in my project, however the same issue can be replicated with the `translate` directive.

In the component spec, try:

```
it('supports conditional rendering such as v-if, v-else-if, v-else', (done) => {
    Vue.config.language = 'en_US'
    let vm = new Vue({
      template: `
      <translate v-if="show">Pending</translate>
      <translate v-else>Hello %{ name }</translate>
      `,
      data: {show: true, name: 'John Doe'},
    }).$mount()
    expect(vm.$el.innerHTML).to.equal('Pending')
    console.log(vm.$el)
    vm.show = false
    vm.$nextTick(function () {
      expect(vm.$el.innerHTML).to.equal('Hello John Doe')
      done()
    })
  })
```

It will output:

```
ERROR: AssertionError{message: 'expected 'Pending' to equal 'Hello John Doe'', showDiff: true, actual: 'Pending', expected: 'Hello John Doe',
```

## Root cause

It is because of:

> The key special attribute is primarily used as a hint for Vue’s virtual DOM algorithm to identify VNodes when diffing ... Vue uses an algorithm that minimizes element movement and tries to patch/reuse elements of the same type in-place as much as possible. 

as explained in the section of [special attribute 'key'](https://vuejs.org/v2/api/#key)

## Solution 1

Add `key` attribute manually to `translate` or `v-translate`

```
<translate key="1" v-if="show">Pending</translate>
<translate key="2" v-else>Hello %{ name }</translate>
```

**This is the immediate fix for anyone having this issue.**

However this is not really ideal because we need to ensure that the keys are unique within each `v-if` block.

## Solution 2

Improve `translate` and `v-translate` to auto-generate `key` attribute if it's not present. I'll be looking at this, need to find a good way to generate unique keys.

Any idea?

**Comment by kemar** [comment]
@kennyki excellent work, thanx!

A good way to generate a unique key is to use a `uuid`, something like that https://gist.github.com/jcxplorer/823878

**Comment by kennyki** [comment]
I have created a separate issue about this matter with an immediate fix available. Please have a look at #29

**Comment by kennyki** [comment]
Oops sorry @kemar I thought that's it's more organized to have a separate issue

**Comment by kemar** [comment]
Problems with `v-if` have been identified and fixed in [v2.0.16](https://github.com/Polyconseil/vue-gettext/releases/tag/v2.0.16).
